In [1]:
import pandas as pd

print("Notebook Ready")

Notebook Ready


# Retail Sales Data Processing and Business Insights Generation

## Data Ingestion and Understanding

Objective:
The objective of this project is to build an end-to-end data engineering pipeline that ingests, cleans, transforms, validates, and analyzes retail sales data to generate business insights and Power BI dashboards.

Datasets Used:
1. product_details
2. retail_data1
3. retail_data2

In [3]:
import pandas as pd

file_path = "../Dataset/Data.xlsx"

retail1 = pd.read_excel(file_path, sheet_name="retail_data1")
retail2 = pd.read_excel(file_path, sheet_name="retail_data2")
products = pd.read_excel(file_path, sheet_name="product_details")

print("Files Loaded Successfully")

Files Loaded Successfully


In [4]:
print("Retail Data 1 Shape:", retail1.shape)
print("Retail Data 2 Shape:", retail2.shape)
print("Product Details Shape:", products.shape)

Retail Data 1 Shape: (4243, 16)
Retail Data 2 Shape: (4251, 16)
Product Details Shape: (10, 4)


In [5]:
retail1.head()

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
0,1,642,Troy Mitchell,108,15000.0,Mixer grinder,Home Appliances,offline,Bangalore,2025-12-06 00:00:00,3,UPI,0.10,Troy60@gmail.com,8385276968,successful
1,2,881,Sarah Guerrero,102,70000.0,Phone,ELEC,online,Bangalore,2025-07-21 00:00:00,5,Card,0.35,SarahGuerrero@yahoo.com,7147248911,successful
2,3,505,Samantha Hull,109,65000.0,Refrigerator,Home Appliances,offline,Chennai,2025-07-11 00:00:00,2,UPI,0.05,SamanthaHull@outlook.com,7415321565,successful
3,4,794,Gerald Cooper,106,80000.0,sofa,FURN,offline,Chennai,2025-10-06 00:00:00,2,UPI,0.05,GeraldCooper@gmail.com,9739354201,successful
4,5,864,Cameron Black,108,15000.0,Mixer grinder,home appliances,offline,Chennai,2025-12-12 00:00:00,1,Cash,0.05,CameronBlack@yahoo.com,9938169477,successful


In [6]:
retail2.head()

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
0,4001,1112,Jasmine Carlson,106,80000.0,sofa,Furniture,offline,Delhi,12-18-2025,2,Cash,0.10,JasmineCarlson@gmail.com,8286662057,successful
1,4002,1125,Jeanne Suarez,105,45000.0,Tv,Electronics,online,Mumbai,2025-08-07 00:00:00,3,Card,0.10,JeanneSuarez@yahoo.com,6435924212,successful
2,4003,1379,Anthony Howard,107,50000.0,Dining table,FURN,offline,Delhi,09-19-2025,2,Cash,0.30,Anthony42@gmail.com,8187928333,successful
3,4004,1570,John Wilkins,107,50000.0,DINING TABLE,FURN,offline,Delhi,2025-11-20 00:00:00,2,Cash,0.40,JohnWilkins@outlook.com,6440039629,successful
4,4005,1433,Joseph Knapp,108,15000.0,Mixer grinder,home appliances,online,Bangalore,2025-09-29 00:00:00,5,Cash,0.25,JosephKnapp@gmail.com,9284221237,successful


In [7]:
products.head()

,product_id,product_name,category,price
0,101,Laptop,Electronics,250000
1,102,Phone,Electronics,70000
2,103,Shirt,Clothing,1300
3,104,Shoes,Clothing,8000
4,105,TV,Electronics,45000


## Initial Dataset Observations

- Three datasets were provided.
- Retail datasets contain transaction-level sales information.
- Product dataset contains master product information.
- Retail datasets contain customer information, pricing, discounts, locations, and payment details.
- Product dataset will be used as a reference table for validation and enrichment.

In [8]:
print("Retail Data 1 Columns")
print(retail1.columns.tolist())

print("\nRetail Data 2 Columns")
print(retail2.columns.tolist())

print("\nProduct Details Columns")
print(products.columns.tolist())

Retail Data 1 Columns
['transaction_id', 'customer_id', 'customer_name', 'product_id', 'price', 'product_name', 'category', 'purchase_location', 'city', 'transaction_date', 'quantity', 'payment_method', 'discount', 'email', 'phone', 'payment_status']

Retail Data 2 Columns
['transaction_id', 'customer_id', 'customer_name', 'product_id', 'price', 'product_name', 'category', 'purchase_location', 'city', 'transaction_date', 'quantity', 'payment_method', 'discount', 'email', 'phone', 'payment_status']

Product Details Columns
['product_id', 'product_name', 'category', 'price']


# Data Profiling and Data Quality Assessment

The datasets were analyzed to identify data quality issues such as:

- Missing values
- Duplicate records
- Invalid values
- Inconsistent formats
- Data type mismatches
- PII fields requiring masking

This assessment helps define the cleaning and transformation strategy.

In [9]:
print("Retail Data 1 Information")
retail1.info()

print("\nRetail Data 2 Information")
retail2.info()

print("\nProduct Details Information")
products.info()

Retail Data 1 Information
<class 'pandas.DataFrame'>
RangeIndex: 4243 entries, 0 to 4242
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   transaction_id     4243 non-null   int64  
 1   customer_id        4243 non-null   int64  
 2   customer_name      4243 non-null   str    
 3   product_id         4243 non-null   int64  
 4   price              3839 non-null   float64
 5   product_name       4243 non-null   str    
 6   category           4243 non-null   str    
 7   purchase_location  4243 non-null   str    
 8   city               4243 non-null   str    
 9   transaction_date   4243 non-null   object 
 10  quantity           4243 non-null   int64  
 11  payment_method     4243 non-null   str    
 12  discount           4243 non-null   float64
 13  email              4243 non-null   str    
 14  phone              4243 non-null   int64  
 15  payment_status     4243 non-null   str    
dtypes: float6

In [10]:
print("Missing Values in Retail Data 1")
print(retail1.isnull().sum())

print("\nMissing Values in Retail Data 2")
print(retail2.isnull().sum())

print("\nMissing Values in Product Details")
print(products.isnull().sum())

Missing Values in Retail Data 1
transaction_id         0
customer_id            0
customer_name          0
product_id             0
price                404
product_name           0
category               0
purchase_location      0
city                   0
transaction_date       0
quantity               0
payment_method         0
discount               0
email                  0
phone                  0
payment_status         0
dtype: int64

Missing Values in Retail Data 2
transaction_id         0
customer_id            0
customer_name          0
product_id             0
price                405
product_name           0
category               0
purchase_location      0
city                   0
transaction_date       0
quantity               0
payment_method         0
discount               0
email                  0
phone                  0
payment_status         0
dtype: int64

Missing Values in Product Details
product_id      0
product_name    0
category        0
price           0
dt

In [11]:
print("Duplicates in Retail Data 1:",
      retail1.duplicated().sum())

print("Duplicates in Retail Data 2:",
      retail2.duplicated().sum())

print("Duplicates in Product Details:",
      products.duplicated().sum())

Duplicates in Retail Data 1: 0
Duplicates in Retail Data 2: 0
Duplicates in Product Details: 0


In [12]:
print("Retail Data 1 Categories")
print(retail1["category"].unique())

print("\nRetail Data 2 Categories")
print(retail2["category"].unique())

Retail Data 1 Categories
<StringArray>
['Home Appliances',            'ELEC',            'FURN', 'home appliances',
     'Electronics',     'electronics',           'CLOTH',        'clothing',
        'Clothing',       'furniture',       'Furniture',            'HOME']
Length: 12, dtype: str

Retail Data 2 Categories
<StringArray>
[      'Furniture',     'Electronics',            'FURN', 'home appliances',
     'electronics',            'HOME',           'CLOTH', 'Home Appliances',
       'furniture',        'Clothing',            'ELEC',        'clothing']
Length: 12, dtype: str


In [13]:
print(retail1["transaction_date"].head(20))

0     2025-12-06 00:00:00
1     2025-07-21 00:00:00
2     2025-07-11 00:00:00
3     2025-10-06 00:00:00
4     2025-12-12 00:00:00
5     2025-07-25 00:00:00
6     2025-07-25 00:00:00
7     2026-03-03 00:00:00
8              02-19-2026
9     2025-08-30 00:00:00
10    2025-12-16 00:00:00
11    2025-12-15 00:00:00
12             06-30-2025
13    2025-06-14 00:00:00
14    2026-02-24 00:00:00
15    2026-01-31 00:00:00
16             08-20-2025
17    2025-05-19 00:00:00
18             05-15-2025
19    2025-09-07 00:00:00
Name: transaction_date, dtype: object


In [14]:
print(retail1["quantity"].describe())
print(retail2["quantity"].describe())

count    4243.000000
mean        2.936130
std         1.474096
min        -2.000000
25%         2.000000
50%         3.000000
75%         4.000000
max         5.000000
Name: quantity, dtype: float64
count    4251.000000
mean        2.942837
std         1.458211
min        -2.000000
25%         2.000000
50%         3.000000
75%         4.000000
max         5.000000
Name: quantity, dtype: float64


## PII Identification

The following columns contain Personally Identifiable Information (PII):

- customer_name
- email
- phone

These fields will be masked during the transformation phase to ensure data privacy and governance compliance.

## Data Quality Findings

The profiling phase identified the following issues:

### Missing Values
- Retail Data 1 contains 404 missing values in the price column.
- Retail Data 2 contains 405 missing values in the price column.

### Category Inconsistency
The category field contains inconsistent values such as:
- Electronics, electronics, ELEC
- Furniture, furniture, FURN
- Home Appliances, home appliances, HOME
- Clothing, clothing, CLOTH

### Date Format Issues
The transaction_date column contains mixed date formats.

Examples:
- 2025-12-06
- 02-19-2026
- 06-30-2025

### Invalid Quantity Values
Negative quantities were identified.
Minimum quantity observed = -2

### PII Fields
The following fields contain sensitive customer information:
- customer_name
- email
- phone

These fields require masking before reporting.

# Data Cleaning and Transformation

In [15]:
retail_data = pd.concat(
    [retail1, retail2],
    ignore_index=True
)

print("Combined Dataset Shape:")
print(retail_data.shape)

Combined Dataset Shape:
(8494, 16)


In [16]:
print(
    retail_data["price"]
    .isnull()
    .sum()
)

809


In [17]:
price_lookup = (
    products
    .set_index("product_id")["price"]
    .to_dict()
)

In [18]:
retail_data["price"] = retail_data.apply(
    lambda row:
    price_lookup.get(
        row["product_id"],
        row["price"]
    )
    if pd.isna(row["price"])
    else row["price"],
    axis=1
)

In [19]:
retail_data["price"].isnull().sum()

np.int64(0)

In [20]:
retail_data["category"] = (
    retail_data["category"]
    .replace({
        "ELEC": "Electronics",
        "electronics": "Electronics",

        "FURN": "Furniture",
        "furniture": "Furniture",

        "HOME": "Home Appliances",
        "home appliances": "Home Appliances",

        "CLOTH": "Clothing",
        "clothing": "Clothing"
    })
)

In [21]:
print(
    retail_data["category"]
    .unique()
)

<StringArray>
['Home Appliances', 'Electronics', 'Furniture', 'Clothing']
Length: 4, dtype: str


In [22]:
retail_data["transaction_date"] = pd.to_datetime(
    retail_data["transaction_date"],
    errors="coerce"
)

In [23]:
retail_data["transaction_date"].head()

0   2025-12-06
1   2025-07-21
2   2025-07-11
3   2025-10-06
4   2025-12-12
Name: transaction_date, dtype: datetime64[us]

In [24]:
(
retail_data["quantity"] < 0
).sum()

np.int64(65)

## Quantity Validation

A total of 65 records contained negative quantities.

Since retail sales cannot have negative quantities, these records were classified as invalid transactions and removed from the curated dataset.

In [25]:
invalid_quantity_count = (
    retail_data["quantity"] < 0
).sum()

print("Invalid Records:", invalid_quantity_count)

retail_data = retail_data[
    retail_data["quantity"] > 0
]

print("New Shape:")
print(retail_data.shape)

Invalid Records: 65
New Shape:
(8401, 16)


## PII Protection

Personally Identifiable Information (PII) fields were masked to protect customer privacy.

Fields masked:
- customer_name
- email
- phone

In [26]:
def mask_name(name):
    if pd.isna(name):
        return name

    return name[0] + "***"

In [27]:
retail_data["customer_name"] = (
    retail_data["customer_name"]
    .apply(mask_name)
)

In [28]:
def mask_email(email):
    if pd.isna(email):
        return email

    username, domain = email.split("@")

    return username[:2] + "***@" + domain

In [29]:
retail_data["email"] = (
    retail_data["email"]
    .apply(mask_email)
)

In [30]:
def mask_phone(phone):
    phone = str(phone)

    return "******" + phone[-4:]

In [31]:
retail_data["phone"] = (
    retail_data["phone"]
    .apply(mask_phone)
)

In [32]:
retail_data[
[
"customer_name",
"email",
"phone"
]
].head()

,customer_name,email,phone
0,T***,Tr***@gmail.com,******6968
1,S***,Sa***@yahoo.com,******8911
2,S***,Sa***@outlook.com,******1565
3,G***,Ge***@gmail.com,******4201
4,C***,Ca***@yahoo.com,******9477


## Revenue Calculation

Revenue was calculated using:

Revenue = Quantity × Price × (1 - Discount)

In [33]:
retail_data["revenue"] = (
    retail_data["quantity"]
    *
    retail_data["price"]
    *
    (1 - retail_data["discount"])
)

In [34]:
retail_data[
[
"quantity",
"price",
"discount",
"revenue"
]
].head()

,quantity,price,discount,revenue
0,3,15000.0,0.10,40500.0
1,5,70000.0,0.35,227500.0
2,2,65000.0,0.05,123500.0
3,2,80000.0,0.05,152000.0
4,1,15000.0,0.05,14250.0


In [35]:
retail_data["year"] = (
    retail_data["transaction_date"]
    .dt.year
)

retail_data["month"] = (
    retail_data["transaction_date"]
    .dt.month_name()
)

retail_data["quarter"] = (
    retail_data["transaction_date"]
    .dt.quarter
)

In [36]:
(retail_data["quantity"] <= 0).sum()

np.int64(0)

In [37]:
retail_data["quantity"].describe()

count    8401.000000
mean        2.983811
std         1.409399
min         1.000000
25%         2.000000
50%         3.000000
75%         4.000000
max         5.000000
Name: quantity, dtype: float64

# Business KPI Generation

After cleaning and transformation, business KPIs were generated to support reporting and decision-making.

In [38]:
total_revenue = retail_data["revenue"].sum()

print("Total Revenue:")
print(round(total_revenue, 2))

Total Revenue:
1239132235.0


In [39]:
total_orders = retail_data["transaction_id"].nunique()

print("Total Orders:")
print(total_orders)

Total Orders:
7914


In [40]:
total_customers = retail_data["customer_id"].nunique()

print("Total Customers:")
print(total_customers)

Total Customers:
1960


In [41]:
average_order_value = (
    retail_data["revenue"].sum()
    /
    retail_data["transaction_id"].nunique()
)

print("Average Order Value:")
print(round(average_order_value, 2))

Average Order Value:
156574.71


In [42]:
revenue_by_category = (
    retail_data
    .groupby("category")["revenue"]
    .sum()
    .sort_values(ascending=False)
)

revenue_by_category

category
Electronics        721682500.0
Furniture          257064000.0
Home Appliances    242328000.0
Clothing            18057735.0
Name: revenue, dtype: float64

In [43]:
revenue_by_city = (
    retail_data
    .groupby("city")["revenue"]
    .sum()
    .sort_values(ascending=False)
)

revenue_by_city.head(10)

city
Chennai      259249060.0
Delhi        256872880.0
Hyderabad    248807740.0
Mumbai       247071395.0
Bangalore    227131160.0
Name: revenue, dtype: float64

In [44]:
top_products = (
    retail_data
    .groupby("product_name")["revenue"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

top_products

product_name
LAPTOP          173425000.0
Laptop          164100000.0
laptop          163150000.0
Sofa             55332000.0
SOFA             52132000.0
sofa             50780000.0
PHONE            46473000.0
Phone            46389000.0
refrigerator     45074250.0
phone            41898500.0
Name: revenue, dtype: float64

In [45]:
monthly_revenue = (
    retail_data
    .groupby("month")["revenue"]
    .sum()
)

monthly_revenue

month
April        101692490.0
August       110027100.0
December      97165725.0
February     106016745.0
January      102941900.0
July         106313690.0
June         105366180.0
March         95669240.0
May          102288240.0
November      99056545.0
October      107293285.0
September    105301095.0
Name: revenue, dtype: float64

## Product Name Standardization

Product names contained inconsistent capitalization patterns.

Examples:
- Laptop, LAPTOP, laptop
- Sofa, SOFA, sofa
- Phone, PHONE, phone

To ensure accurate reporting and aggregation, product names were standardized using title case formatting.

In [46]:
retail_data["product_name"] = (
    retail_data["product_name"]
    .str.strip()
    .str.title()
)

In [47]:
top_products = (
    retail_data
    .groupby("product_name")["revenue"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

top_products

product_name
Laptop           500675000.0
Sofa             158244000.0
Phone            134760500.0
Refrigerator     124390500.0
Dining Table      98820000.0
Microwave         89214750.0
Tv                86247000.0
Mixer Grinder     28722750.0
Shoes             15632000.0
Shirt              2425735.0
Name: revenue, dtype: float64

# Final Curated Dataset

The dataset has now undergone:

- Missing value treatment
- Category standardization
- Product standardization
- Date normalization
- Invalid quantity removal
- PII masking
- Revenue calculation
- Time attribute creation

The resulting dataset is ready for business intelligence reporting and dashboarding.

In [48]:
retail_data.to_csv(
    "../Dataset/final_cleaned_data.csv",
    index=False
)

print("Final dataset exported successfully.")

Final dataset exported successfully.


In [49]:
data_quality_summary = pd.DataFrame({
    "Issue": [
        "Missing Prices",
        "Category Inconsistency",
        "Mixed Date Formats",
        "Negative Quantities",
        "PII Exposure"
    ],
    "Resolution": [
        "Filled using product master",
        "Standardized categories",
        "Converted to datetime",
        "Removed invalid rows",
        "Applied masking"
    ]
})

data_quality_summary

,Issue,Resolution
0,Missing Prices,Filled using product master
1,Category Inconsistency,Standardized categories
2,Mixed Date Formats,Converted to datetime
3,Negative Quantities,Removed invalid rows
4,PII Exposure,Applied masking


In [50]:
retail_data["product_name"] = (
    retail_data["product_name"]
    .str.strip()
    .str.title()
)

top_products = (
    retail_data
    .groupby("product_name")["revenue"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

top_products

product_name
Laptop           500675000.0
Sofa             158244000.0
Phone            134760500.0
Refrigerator     124390500.0
Dining Table      98820000.0
Microwave         89214750.0
Tv                86247000.0
Mixer Grinder     28722750.0
Shoes             15632000.0
Shirt              2425735.0
Name: revenue, dtype: float64